go/escaper/escaper.go

```diff
@@ -22,10 +22,27 @@ func CleanBytes(p []byte) []byte {
 	return []byte(Clean(string(p)))
 }
 
+// Writer can be used to write data to the underlying io.Writer, while transparently sanitizing it.
+// If an error occurs writing to a Writer, all subsequent writes will return the error.
+// Note that the sanitization might alter the size of the actual data being written.
 type Writer struct {
+	err error
 	io.Writer
 }
 
-func (w Writer) Write(p []byte) (int, error) {
-	return w.Writer.Write(CleanBytes(p))
+// Write writes p to the underlying io.Writer, after sanitizing it.
+// It returns n = len(p) on a successful write (regardless of how much data is written).
+// This is because the escaping function might alter the actual dimension of the data, but the caller is interested
+// in knowing how much of what they wanted to write was actually written. In case of errors it (conservatively) returns n=0
+// and the error, and no other writes are possible.
+func (w *Writer) Write(p []byte) (int, error) {
+	if w.err != nil {
+		return 0, w.err
+	}
+	if _, err := w.Writer.Write(CleanBytes(p)); err == nil {
```

Comment:
remind me: is it guaranteed there won't be any "short writes"?

Reply:
Not sure about what you're asking here. 
If you are asking about why the function returns 0 on errors, my reasoning is that because of escaping the amount of bytes written to the stream and returned by w.Writer.Write(...) does not map directly to the number of bytes the caller intended to write. 
I could have written some more code to try and figure out how many bytes of the original stream were written, but this can be problematic if the escaping substitution we are doing changes the number of bytes of the symbols, and it is probably not useful anyways (I could not find any logic to retry writes, but I might be wrong on this). So I decided it was just easier to return 0 and document that. Happy to change that if you disagree.

Also, since it can be non trivial/error prone to recover from a failed write in this case (i.e. trying again), I decided to just prevent any further writing after an error. This behavior (locking further writes on error) is consistent with what bufio.Writer does.

Reply:
My question isn’t that. You are throwing away the first return value from write. Traditionally that is because you can pass write n bytes and it might only write out m bytes, where m<n.  In that case you need to rewrite with the remainder. This is called a “short write.”  Is that a possibility here?

Reply:
My question isn’t that. You are throwing away the first return value from write. Traditionally that is because you can pass write n bytes and it might only write out m bytes, where m<n.  In that case you need to rewrite with the remainder. This is called a “short write.”  Is that a possibility here?

Reply:
No, that is not a guarantee. In that situation Write would also have to return an error (that's mandated in the definition of the io.Writer interface), in which case this function returns the error and prevents further writes (rather than retrying). 
Are we doing retries on short writes somewhere else in the code that this can mess up with?

Reply:
If you get an error back from write then it is fine to just error all the way out. Thanks.